File: supabase/functions/calculate-risk/services/elevation.ts
const ELEVATION_API = 'https://api.open-meteo.com/v1/elevation';
const TIMEOUT_MS = 3000;

interface OpenMeteoElevationResponse {
  elevation: {
    elevation: number[];
  };
}

const cache = new Map<string, number>();

function cacheKey(lat: number, lon: number): string {
  const k = (v: number) => Math.round(v * 1e4) / 1e4;
  return `${k(lat)},${k(lon)}`;
}

export async function getElevation(
  latitude: number,
  longitude: number,
): Promise<number | null> {
  const key = cacheKey(latitude, longitude);
  const cached = cache.get(key);
  if (cached !== undefined) return cached;

  try {
    const url = `${ELEVATION_API}?latitude=${latitude}&longitude=${longitude}`;
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), TIMEOUT_MS);

    const res = await fetch(url, { signal: controller.signal });
    clearTimeout(timeout);

    if (!res.ok) return null;

    const json: OpenMeteoElevationResponse = await res.json();
    const elevation = json.elevation.elevation[0];

    if (elevation != null) {
      cache.set(key, elevation);
    }

    return elevation ?? null;
  } catch {
    return null;
  }
}

export function clearCache(): void {
  cache.clear();
}

File: bot/src/services/elevation.py
import logging

import httpx

logger = logging.getLogger(__name__)

OPEN_METEO_ELEVATION_URL = "https://api.open-meteo.com/v1/elevation"


async def get_elevation(lat: float, lon: float, timeout: float = 5.0) -> float | None:
    params = {"latitude": lat, "longitude": lon}

    try:
        async with httpx.AsyncClient(timeout=timeout) as client:
            resp = await client.get(OPEN_METEO_ELEVATION_URL, params=params)
            resp.raise_for_status()
            data = resp.json()

            elevation = data["elevation"][0]
            logger.info("Elevation at (%.4f, %.4f): %.1f m", lat, lon, elevation)
            return float(elevation)

    except Exception as e:
        logger.warning("Elevation fetch failed for (%.4f, %.4f): %s", lat, lon, e)
        return None
